proof/refine/ARM/KHeap_R.thy

```diff
@@ -3921,4 +3935,315 @@ lemma threadGet_getObject:
+  apply (rule iffI; subgoal_tac "sc_at' p s \<and> tcb_at' t s", clarsimp)
+     apply (clarsimp simp: sc_tcbs_of_pred_map_equiv tcb_scs_of_pred_map_equiv)
+     apply (clarsimp simp: sqheap_refs_inv_defs tcb_scs_of_pred_map_equiv)
+    apply (frule obj_at'_typ_at', simp)
+    apply (frule obj_at_ko_at', clarsimp)
+    apply (frule (1) sc_ko_at_valid_objs_valid_sc')
+    apply (clarsimp simp: valid_sched_context'_def)
+   apply (clarsimp simp: sc_tcbs_of_pred_map_equiv tcb_scs_of_pred_map_equiv)
+   apply (clarsimp simp: sqheap_refs_inv_defs tcb_scs_of_pred_map_equiv)
+  apply (frule obj_at'_typ_at', simp)
+  apply (frule obj_at_ko_at', clarsimp)
+  apply (frule (1) tcb_ko_at_valid_objs_valid_tcb')
+  apply (clarsimp simp: valid_tcb'_def)
+  done
+
+lemma not_idle_scTCB:
+  "tcbs_scs_sym_refs s \<Longrightarrow>
+   valid_objs' s \<Longrightarrow>
+   valid_idle' s \<Longrightarrow>
+   p \<noteq> idle_sc_ptr \<Longrightarrow>
+   sc_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. scTCB x \<noteq> Some idle_thread_ptr) p s"
+  apply (subgoal_tac "\<not>obj_at' (\<lambda>x. scTCB x = Some idle_thread_ptr) p s")
+   apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def)
+  apply clarsimp
+  apply (subst (asm) tcbs_scs_sym_refs_equiv, clarsimp+)
+  apply (clarsimp simp: valid_idle'_def obj_at'_real_def ko_wp_at'_def idle_tcb'_def)
+  done
+
+lemma not_idle_tcbSC:
+  "tcbs_scs_sym_refs s \<Longrightarrow>
+   valid_objs' s \<Longrightarrow>
+   valid_idle' s \<Longrightarrow>
+   p \<noteq> idle_thread_ptr \<Longrightarrow>
+   tcb_at' p s \<Longrightarrow>
+   obj_at' (\<lambda>x. tcbSchedContext x \<noteq> Some idle_sc_ptr) p s"
+  apply (subgoal_tac "\<not>obj_at' (\<lambda>x. tcbSchedContext x = Some idle_sc_ptr) p s")
+   apply (clarsimp simp: obj_at'_real_def ko_wp_at'_def)
+  apply clarsimp
+  apply (subst (asm) tcbs_scs_sym_refs_equiv[symmetric], clarsimp+)
+   apply (clarsimp simp: valid_idle'_def)
+  apply (clarsimp simp: valid_idle'_def obj_at'_real_def ko_wp_at'_def idle_tcb'_def)
+  done
+
+lemma setObject_tcb_tcbs_of':
+  "\<lbrace>\<lambda>s. P' ((tcbs_of' s)(c \<mapsto> tcb))\<rbrace>
+  setObject c (tcb::tcb)
+  \<lbrace>\<lambda>_ s. P' (tcbs_of' s)\<rbrace>"
+  by (setObject_easy_cases, clarsimp simp: ARM_H.fromPPtr_def)
+
+lemma setObject_other_tcbs_of'[wp]:
+  "setObject c (r::reply) \<lbrace>\<lambda>s. P' (tcbs_of' s)\<rbrace>"
+  "setObject c (e::endpoint) \<lbrace>\<lambda>s. P' (tcbs_of' s)\<rbrace>"
+  "setObject c (n::notification) \<lbrace>\<lambda>s. P' (tcbs_of' s)\<rbrace>"
+  "setObject c (sc::sched_context) \<lbrace>\<lambda>s. P' (tcbs_of' s)\<rbrace>"
+  by setObject_easy_cases+
+
+lemma setObject_cte_tcb_scs_of'[wp]:
+  "setObject c (reply::cte) \<lbrace>\<lambda>s. P' (tcb_scs_of' s)\<rbrace>"
+  by setObject_easy_cases
+
+lemma setObject_tcb_scs_of'[wp]:
+  "setObject c (tcb::tcb) \<lbrace>\<lambda>s. P' (scs_of' s)\<rbrace>"
```

Comment:
Is it worth doing this for more object types, like you did for `tcbs_of'`? Or is this a case of only adding different types as necessary.